File: java/MergeTwoSortedLists.java
// 21. Merge Two Sorted Lists
// Accepted 16ms

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode mergeTwoLists(ListNode l1, ListNode l2) {
        ListNode ret = new ListNode(-1);
        ListNode head = ret;
        
        while(l1 != null || l2 != null) {
            
            if(l1 != null && l2 != null) {
                if(l1.val < l2.val) {
                    ListNode newNode = new ListNode(l1.val);
                    newNode.next = null;
                    ret.next = newNode;
                    
                    l1 = l1.next;
                }
                else if(l1.val > l2.val) {
                    ListNode newNode = new ListNode(l2.val);
                    newNode.next = null;
                    ret.next = newNode;
                    
                    l2 = l2.next;
                }
                else {
                    ListNode newNode = new ListNode(l1.val);
                    newNode.next = null;
                    ret.next = newNode;
                    
                    ret = ret.next;
                    l1 = l1.next;
                    
                    newNode = new ListNode(l2.val);
                    newNode.next = null;
                    ret.next = newNode;
                    
                    l2 = l2.next;
                }
            }
            
            else if(l1 != null) {
                ListNode newNode = new ListNode(l1.val);
                newNode.next = null;
                ret.next = newNode;

                l1 = l1.next;
            }
            
            else {
                ListNode newNode = new ListNode(l2.val);
                newNode.next = null;
                ret.next = newNode;

                l2 = l2.next;
            }
            
            ret = ret.next;
        }
        
        return head.next;
    }
}

File: python3/021. Merge Two Sorted Lists/MergeTwoSortedLists.py
# 21. Merge Two Sorted Lists
# Accepted 44ms

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def mergeTwoLists(self, l1, l2):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        
        ret = ListNode(-1)
        retHead = ret
        
        while l1 != None and l2 != None:
            if l1.val < l2.val:
                newNode = ListNode(l1.val)
                l1 = l1.next
            else:
                newNode = ListNode(l2.val)
                l2 = l2.next
            ret.next = newNode; ret = ret.next
        
        while l1 != None:
            ret.next = ListNode(l1.val); ret = ret.next
            l1 = l1.next
        while l2 != None:
            ret.next = ListNode(l2.val); ret = ret.next
            l2 = l2.next
        
        return retHead.next
